content/docs/testing-recipes.md

```diff
@@ -0,0 +1,358 @@
+---
+id: testing-recipes
+title: Test recipes 
+permalink: docs/testing-recipes.html
+---
+
+This document goes through common testing patterns for React components.
+
+- This assumes tests are running in jest. For details on setting up a testing environment, read [the Environments doc](/docs/testing-environments.html).
+- This guide uses React's [`act()`](/docs/act.html) testing helper.
+
+### Rendering {#rendering}
+
+Testing whether a component renders correctly for given props is a useful signal to the stability of a component. Consider a simple component that renders a message based on a prop.
+
+```jsx
+// hello.js
+export default function Hello(props) {
+  return props.name === undefined ? (
+    <h1>Hello, {props.name}!</h1>
+  ) : (
+    <span>Hey, stranger</span>
+  );
+}
+```
+
+We can write a test for this component:
+
+```jsx
+// hello.test.js
+import Hello from "./hello";
+import { render } from "react-dom";
+import { act } from "react-dom/test-utils";
+
+it("renders with or without a name", () => {
+  const container = document.createElement("div");
+  act(() => {
+    render(<Hello />, container);
+  });
+  expect(container.textContent).toBe("Hey, stranger");
+
+  act(() => {
+    render(<Hello name="Sophie" />, container);
+  });
+  expect(container.textContent).toBe("Hello Sophie!");
+
+  act(() => {
+    render(<Hello name="Flarnie" />, container);
+  });
+  expect(container.textContent).toBe("Hello Flarnie!");
+  // cleanup on exiting
+  container.remove();
```

Comment:
This seems like an anti pattern since it won’t run on failure.

Reply:
added setup/teardown for all the tests